flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/state/MultiJoinStateViews.java

```diff
@@ -0,0 +1,440 @@
+            reusedList.clear();
+            RowData record = recordState.get(joinKey);
+            if (record != null) {
+                reusedList.add(record);
+            }
+            return reusedList;
+        }
+
+        @Override
+        public void cleanup(RowData joinKey) throws Exception {
+            recordState.remove(joinKey);
+        }
+    }
+
+    /**
+     * State view for input sides that have a unique key, but it differs from the join key.
+     *
+     * <p>Stores data as {@code MapState<CompositeKey<JoinKey, UK>, Record>}. The composite key is a
+     * RowData with 2 fields: joinKey and uniqueKey.
+     */
+    private static final class InputSideHasUniqueKey implements MultiJoinStateView {
+
+        // stores record in the mapping <(JoinKey, UK), Record>
+        private final MapState<RowData, RowData> recordState;
+        private final KeySelector<RowData, RowData> uniqueKeySelector;
+        private final InternalTypeInfo<RowData> joinKeyType;
+
+        private InputSideHasUniqueKey(
+                RuntimeContext ctx,
+                final String stateName,
+                final InternalTypeInfo<RowData> joinKeyType,
+                final InternalTypeInfo<RowData> recordType,
+                final InternalTypeInfo<RowData> uniqueKeyType,
+                final KeySelector<RowData, RowData> uniqueKeySelector,
+                final StateTtlConfig ttlConfig) {
+            checkNotNull(uniqueKeyType);
+            checkNotNull(uniqueKeySelector);
+            this.uniqueKeySelector = uniqueKeySelector;
+            this.joinKeyType = joinKeyType;
+
+            // Composite key type: RowData with 2 fields (joinKey, uniqueKey)
+            // The composite key is a RowData with joinKey at index 0 and uniqueKey at index 1.
+            // TODO Gustavo: there is probably a cleaner way of instantiating the type
+            final RowType keyRowType =
+                    RowType.of(joinKeyType.toRowType(), uniqueKeyType.toRowType());
+            InternalTypeInfo<RowData> compositeKeyType = InternalTypeInfo.of(keyRowType);
+
+            MapStateDescriptor<RowData, RowData> recordStateDesc =
+                    createStateDescriptor(stateName, compositeKeyType, recordType, ttlConfig);
+
+            this.recordState = ctx.getMapState(recordStateDesc);
+        }
+
+        private RowData createCompositeKey(RowData joinKey, RowData uniqueKey) {
+            GenericRowData compositeKey = new GenericRowData(2);
+            compositeKey.setField(0, joinKey);
+            compositeKey.setField(1, uniqueKey);
+            return compositeKey;
+        }
+
+        @Override
+        public void addRecord(RowData joinKey, RowData record) throws Exception {
+            RowData uniqueKey = uniqueKeySelector.getKey(record);
+            RowData compositeKey = createCompositeKey(joinKey, uniqueKey);
+            recordState.put(compositeKey, record);
```

Comment:
Should we project the input record and only store the needed fields to reduce space consumption and IO?

Reply:
That's an interesting idea. Currently, a StreamingJoinOperator is locally chained to a Calc inside of a Task which is kind of a similar optimization. Your suggestion is basically that we merge both operators and that the MJOperator takes care of materializing (push down the optimization), if I understand it correctly. 

In theory it makes sense. From an optimizer perspective we'd definitely need a couple of changes. I wonder why the other operators do not do this - maybe there's a drawback/reason behind it? I'd either way suggest addressing this as part of another ticket since it's a separate optimization. But I think it's something we should look into.

Reply:
I think it's all about priorities, I know a couple of other places where projections could be pushed down as well, but  so far was not prioritized or was addressed differently.

However, in this case it **_might_** be more pressing because the amount of state to process a single record might be bigger.

Addressing this in a separate ticket (if we decide to do so) would make sense, but we need to keep in mind that the state would have to be migrated (or we support this only for the new jobs somehow).

Reply:
> However, in this case it might be more pressing because the amount of state to process a single record might be bigger.

There will be N states for N inputs, and the number of reads might grow exponentially (depending on the "branching" factor, i.e. how many matches do we have for a single record).
That might exhaust RocksDB block cache (especially given that it doesn't separate keys and values).
So at a first glance, this optimization also seems to be highly desirable (but again, good to have some benchmark results).